packages/fx-core/src/plugins/resource/keyvault/utils/telemetry.ts

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+import { PluginContext } from "@microsoft/teamsfx-api";
+import { Constants, Telemetry } from "../constants";
+
+export class TelemetryUtils {
+  static ctx: PluginContext;
+
+  public static init(ctx: PluginContext) {
+    TelemetryUtils.ctx = ctx;
+  }
+
+  public static sendEvent(
+    eventName: string,
+    properties?: { [key: string]: string },
+    measurements?: { [key: string]: number }
+  ) {
+    if (!properties) {
+      properties = {};
+    }
+    properties[Telemetry.isSuccess] = Telemetry.success;
+    properties[Telemetry.component] = Constants.KeyVaultPlugin.pluginName;
+    TelemetryUtils.addAppIdInProperty(properties, this.ctx);
+    TelemetryUtils.ctx.telemetryReporter?.sendTelemetryEvent(eventName, properties, measurements);
+  }
+
+  public static sendErrorEvent(
+    eventName: string,
+    errorName: string,
+    errorType: string,
+    errorMessage: string,
+    properties?: { [key: string]: string },
+    measurements?: { [key: string]: number }
+  ) {
+    if (!properties) {
+      properties = {};
+    }
+    properties[Telemetry.isSuccess] = Telemetry.fail;
+    properties[Telemetry.component] = Constants.KeyVaultPlugin.pluginName;
+    properties[Telemetry.errorCode] = `${Constants.KeyVaultPlugin.shortName}.${errorName}`;
+    properties[Telemetry.errorType] = errorType;
+    properties[Telemetry.errorMessage] = errorMessage;
+    TelemetryUtils.addAppIdInProperty(properties, this.ctx);
+    TelemetryUtils.ctx.telemetryReporter?.sendTelemetryErrorEvent(
+      eventName,
+      properties,
+      measurements
+    );
+  }
+
+  private static addAppIdInProperty(
```

Comment:
I remember we also have project id, should we also record project id in telemetry? since you only generates arm template, if user didn't provision, there will be no Teams App id.
cc @chagong @KennethBWSong

Reply:
Project id will be added in extension and cli level.

Reply:
Sync with @KennethBWSong : 

> No we don't need to do that since it is provided in cli and extension